internal/state/execution.go

```diff
@@ -141,25 +141,31 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
 		panic(err)
 	}
 
-	if err := rpp.Validate(maxDataBytes, block.Txs.ToSliceOfBytes()); err != nil {
-		return nil, err
-	}
-
 	if !rpp.ModifiedTx {
 		return block, nil
 	}
+	txrSet := types.NewTxRecordSet(rpp.TxRecords)
 
-	for _, rtx := range rpp.RemovedTxs() {
-		if err := blockExec.mempool.RemoveTxByKey(types.Tx(rtx.Tx).Key()); err != nil {
-			blockExec.logger.Debug("error removing transaction from the mempool", "error", err)
+	if err := txrSet.Validate(maxDataBytes, block.Txs); err != nil {
+		return nil, err
+	}
+
+	for _, rtx := range txrSet.GetRemovedTxs() {
+		if err := blockExec.mempool.RemoveTxByKey(rtx.Key()); err != nil {
+			blockExec.logger.Debug("error removing transaction from the mempool", "error", err, "tx hash", rtx.Hash())
 		}
 	}
-	for _, rtx := range rpp.AddedTxs() {
-		if err := blockExec.mempool.CheckTx(ctx, rtx.Tx, nil, mempool.TxInfo{}); err != nil {
-			blockExec.logger.Error("error adding tx to the mempool", "error", err)
+	for _, atx := range txrSet.GetAddedTxs() {
+		if err := blockExec.mempool.CheckTx(ctx, *atx, nil, mempool.TxInfo{}); err != nil {
+			blockExec.logger.Error("error adding tx to the mempool", "error", err, "tx hash", atx.Hash())
 		}
 	}
-	return state.MakeBlock(height, types.TxRecordsToTxs(rpp.IncludedTxs()), commit, evidence, proposerAddr), nil
+	itxs := append(txrSet.GetAddedTxs(), txrSet.GetUnmodifiedTxs()...)
```

Comment:
do you have to call GetAddedTxs both here and on 158? or can you build the first part of this list there?

Reply:
I definitely could do that, at the moment, the Get is basically free so I figured why not just keep the method call for clarity :man_shrugging: